### PR TITLE
incremental training based on the original model in format of binary (not ftz)

### PR DESCRIPTION
**Pull request by ericxsun**
training the classification model or word vector model incrementally.

### CODE REVIEW

**Comment by ericxsun** [comment]
@EdouardGrave Is it available of reviewing this MR for you. Thanks a lot.

**Comment by matanster** [comment]
@ericxsun am I right to assume from the code, that training from an existing model will only train on the new training data supplied? as a form of transfer learning, did this work very well for you in practice?

**Comment by ericxsun** [comment]
@matanster re-training the existing model no matter the supplied training data is new or not. That's ok for transfer learning theoretically, but I did not do any experiments.

**Comment by matanster** [comment]
I wonder if this ever gets merged. Frankly because the classification model is so simple, I think it should be tweaked outside the fasttext codebase, only taking the word embeddings from a fasttext run. It's really a simple machine learning model (even if its C++ code is slightly convoluted for a layman reader of C++ code).

**Comment by mino98** [comment]
@matanster what do you mean with " only taking the word embeddings from a fasttext run". I don't understand how you'd do incremental training that way, can you please elaborate? Thanks.

**Comment by matanster** [comment]
@mino98 you are right, I've been ambiguous there. I meant to say, branching out from a monolith that computes word embeddings and also performs classification → to a practice where you only take the word embeddings from some (unsupervised) run of fasttext, and re-implement the classification bit somewhere else, with any and all tweaks you like.

**Comment by bangj** [comment]
Any update on this?

**Comment by ericxsun** [comment]
@matanster incremental training actually means, having finished training model with data we got before, and retrain the model with newer data we get, **not from scratch**. 

When training word embedding, you could do it from scratch with all data at each time, or just on the new data. For classification, you could train it from scratch with pre-trained word embedding with all data, or only the new one, with no changing of the word embedding. It's all Okay.

**Comment by ericxsun** [comment]
@bangj no more change

**Comment by matanster** [comment]
Cool PR. This will be mainly helpful for minor domain adaptation, I guess, or do you think these algorithms carry any  online learning guarantees?

**Comment by dcsan** [comment]
this would be great to see in a new release!

**Comment by bangj** [comment]
@ericxsun No I mean facebook guys, cc @cpuhrsch @EdouardGrave @piotr-bojanowski @ajoulin

**Comment by mino98** [comment]
It'd be nice to get this change reviewed and merged.

**Comment by Tyrenn** [comment]
Definitely need this

**Comment by BlockheadLS** [comment]
Wow! Niubility!

**Comment by mahdeto** [comment]
Great feature hope we can get this

**Comment by matanster** [comment]
I find that the [readme](https://github.com/ericxsun/fastText/blob/cd57d29966384737c4a7638449af681d915cd921/README.md) is a bit too lean on what this feature really does. Can anticipate user group questions about the newly introduced `-incr` flag as is. Since much of the audience for this library are not real data scientists, may I suggest at least being less terse ― 

It can easily use language to say what incremental training means, in plain English.

**Comment by loretoparisi** [comment]
@matanster was the README merged?

**Comment by Nazaire** [comment]
Looking forward to using this!

**Comment by mino98** [comment]
Thanks for the last set of commits @ericxsun, in particular "prevent name of output and inputMode be the same" was not obvious and caused me some headaches :)

I'd like to point out a small bug, not sure how you suggest to fix it.

The learning rate (`lr`) is not among the model parameters that are saved in the `bin` file by `Args->save()` ([here](https://github.com/facebookresearch/fastText/blob/3e64bf0f5b916532b34be6706c161d7d0a4957a4/src/args.cc#L245)) and loaded back by `Args->load()` ([here](https://github.com/facebookresearch/fastText/blob/3e64bf0f5b916532b34be6706c161d7d0a4957a4/src/args.cc#L261)). 

Thus [this line](https://github.com/facebookresearch/fastText/pull/423/files#diff-0c9c4a3d7c5d19e9fe56d939f0b5d94dR727) from your PR won't work as intended: if you don't specify `-lr` from the CLI, then the default is used (`0.05` from [here](https://github.com/facebookresearch/fastText/blob/3e64bf0f5b916532b34be6706c161d7d0a4957a4/src/args.cc#L20)), regardless of what you've been using in the previous iterations.

Do you want to fix it by patching `args.cc` to save/load `lr` in the `.bin` file?

To summarise:

for the following params, their values from the saved model override those from the CLI:
* `dim`
* `ws`
* `neg`
* `wordNgrams`
* `loss`
* `model`

while for the following parameters, their values from the CLI override those in the saved model:
* `epoch`
* `minCount`
* `bucket`
* `minn`
* `maxn`
* `lrUpdateRate`
* `t`
* `lr` (currently broken, see above)

**Comment by ericxsun** [comment]
@mino98 thanks for your work, I'll check the code, for restoring all the parameters saved in trained model.

**Comment by mino98** [comment]
Thanks @ericxsun, you are still allowing these params to be set from the CLI during updates:
* `dim`
* `loss`
* `model`
I'm not sure about `loss`, but surely the model will break if you change `dim` or `model` between updates. It makes sense to introduce a check that prints and errors and quits if these three are specified at the command line and `incr==True`, don't you think?

Also, another question: I noticed that you added a python binding to load a model from memory bytes. What is the best way in Python to do a first model training, and then to update it? So far i was simply doing things like:
```
import fastText as ft

# train the model with the first document:
model = ft.train_unsupervised(
    input="my-first-document.txt",
    ...
    incr=False
    inputModel=""
)

# save the model:
model.save_model("my-model.bin")

# then, later... update the model with the new documents:
model = ft.train_unsupervised(
    input="my-second-document.txt",
    ...
    incr=True
    inputModel="my-model.bin"
)

# and save the model over the previous:
model.save_model("my-model.bin")
```

Is this what you have in mind? Loading back the model takes some time, so it would make sense to avoid that step.

**Comment by mino98** [comment]
@ericxsun Ok, but how would you use it during update training? Thanks again.

**Comment by ericxsun** [comment]
@mino98 I guess, saving model in memory, loading it from memory and continue to train would be okay. But I have not do any experiments in this way.

**Comment by selenewang** [comment]
Hi, when I tried to retrain the wikipedia word embedding, I got this error. Can anyone help?

$ ./fasttext cbow -input ../dataset/french_text.txt -inputModel ../wiki.fr/wiki.fr.bin -output fr_retrain -dim 300 -epoch 300 -bucket 10000000 -incr
Update args
Load dict from trained model
Load dict from training data
Read 11M words
Number of words:  84575
Number of labels: 0
Merge dict
Assertion failed: (id < size_), function getWord, file src/dictionary.cc, line 139.
Abort trap: 6

**Comment by Lego1st** [comment]
@ericxsun do you have any solution for the Assertion failed that @selenewang met. I have also struggled with this.

**Comment by ericxsun** [comment]
@selenewang @Lego1st  Sorry to hear that. However on my dataset with same command, no errors occur.  `french_text.txt` does not on my hand, such error could not be reproduced.

**Comment by AritzBi** [comment]
I have the following error too:

fasttext: src/dictionary.cc:139: std::__cxx11::string fasttext::Dictionary::getWord(int32_t) const: Assertion `id < size_' failed.

@ericxsun have you been able to locate the error? thank you very much!

**Comment by skalinin** [comment]
Why do we need -incr if fasttext already has --pretrainedVectors method? Or these things have different meaning?

**Comment by jtoy** [comment]
love it, is this getting merged?

**Comment by ericxsun** [comment]
@AritzBi @selenewang @Lego1st please use the latest code.

**Comment by tranhungnghiep** [comment]
Any update on merging this PR? Will it be supported by the python wrapper here: https://github.com/facebookresearch/fastText/tree/master/python?

**Comment by jtoy** [comment]
is facebook not accepting pull requests? this has been here for a long time.

**Comment by RXminuS** [comment]
Cmon guys, any response?

**Comment by bangj** [comment]
Helo @facebookresearch @cpuhrsch @EdouardGrave @piotr-bojanowski @ajoulin

File: src/fasttext.cc [hunk 1]
```diff
@@ -430,7 +458,7 @@ void FastText::predict(
       }
       std::cout << dict_->getLabel(it->second);
       if (print_prob) {
-        std::cout << " " << std::exp(it->first);
+        std::cout << separator << std::exp(it->first);
```
Comment (mhart) [inline src/fasttext.cc]: It looks like you've changed the default separator from a space to a tab – so this would be a breaking change. Might be better to keep the default as a space?

Comment (ericxsun) [reply src/fasttext.cc]: leaving it along, it kept the default value: space

**Comment by loretoparisi** [comment]
@skalinin please read here about the difference between fine tuning a unsupervised model and using pre-trained word vectors for supervised training - https://github.com/facebookresearch/fastText/issues/713#issuecomment-454333999

**Comment by dvbarnes** [comment]
Has anyone been able to use this yet? Seeing as the authors don't have interest in merging this in, is it something worth forking?

File: src/fasttext.cc [hunk 2]
```diff
@@ -647,33 +668,197 @@ void FastText::loadVectors(std::string filename) {
...
-  ifs.close();
 
-  if (args_->pretrainedVectors.size() != 0) {
-    loadVectors(args_->pretrainedVectors);
-  } else {
-    input_ = std::make_shared<Matrix>(dict_->nwords()+args_->bucket, args_->dim);
-    input_->uniform(1.0 / args_->dim);
-  }
+  if (!args.incr) {
+    std::ifstream ifs(args_->input);
```
Comment (dvbarnes) [inline src/fasttext.cc]: Is there a reason you don't use std::ifstream::binary similar to https://github.com/facebookresearch/fastText/blob/1e6637513bdca926e83e540b0a5296dd0878adca/src/fasttext.cc#L180? this causes the data to be read incorrectly without the binary flag.

**Comment by PCChris1995** [comment]
@ericxsun hi when I tried to retrain the wikipedia word embedding, I got this error. Can youhelp me ?thx!
```
./fasttext skipgram -input ../data/search_sequeces.data -inputModel cc.zh.300.bin -output re-trained  -incr
Unknown argument: -inputModel
```

**Comment by ericxsun** [comment]
> @ericxsun hi when I tried to retrain the wikipedia word embedding, I got this error. Can youhelp me ?thx!
> 
> ```
> ./fasttext skipgram -input ../data/search_sequeces.data -inputModel cc.zh.300.bin -output re-trained  -incr
> ```
> 
> ERROR: Unknown argument: -inputModel
> 
> the parm `-inputModel `is not exist, why hanppen this?

Since this MR did not merged, the official code does not support such param. If you want, see [my repo fasttext](https://github.com/ericxsun/fastText). (Clone the repo and make)

```sh
./fasttext supervised
Empty input or output path.

The following arguments are mandatory:
  -input              training file path
  -output             output file path

The following arguments are optional:
  -inputModel         trained model file path (required only for incremental training)
  -verbose            verbosity level [2]
```

**Comment by Nandha1712** [comment]
@ericxsun After performing the incremental training, the model is correctly predicting with high probability for newly given values(provided in incremental training). 
But it does not predict correctly for other values which were working fine with the original model.

**Comment by nazikh123** [comment]
> @ericxsun After performing the incremental training, the model is correctly predicting with high probability for newly given values(provided in incremental training).
> But it does not predict correctly for other values which were working fine with the original model.

I am also facing the same issue. Any recommendations how to resolve this?

**Comment by ericxsun** [comment]
> > @ericxsun After performing the incremental training, the model is correctly predicting with high probability for newly given values(provided in incremental training).
> > But it does not predict correctly for other values which were working fine with the original model.
> 
> I am also facing the same issue. Any recommendations how to resolve this?

I'll check it and wish to solve such problem

**Comment by mustfkeskin** [comment]
Any progress here

**Comment by eduamf** [comment]
> @ericxsun I wonder if this ever gets merged. Frankly because the classification model is so simple, I think it could equally be tweaked outside the fasttext codebase, only taking the word embeddings from a fasttext run. It's really a simple machine learning model (even if its C++ code implementation is slightly convoluted for a layman reader of C++ code).

Wonderful! 

> > @ericxsun hi when I tried to retrain the wikipedia word embedding, I got this error. Can youhelp me ?thx!
> > ```
> > ./fasttext skipgram -input ../data/search_sequeces.data -inputModel cc.zh.300.bin -output re-trained  -incr
> > ```
> > 
> > 
> > ERROR: Unknown argument: -inputModel
> > the parm `-inputModel `is not exist, why hanppen this?
> 
> Since this MR did not merged, the official code does not support such param. If you want, see [my repo fasttext](https://github.com/ericxsun/fastText). (Clone the repo and make)
> 
> ```shell
> ./fasttext supervised
> Empty input or output path.
> 
> The following arguments are mandatory:
>   -input              training file path
>   -output             output file path
> 
> The following arguments are optional:
>   -inputModel         trained model file path (required only for incremental training)
>   -verbose            verbosity level [2]
> ```

It doesn`t make sense. You did a really good job.

**Comment by zafercavdar** [comment]
Sad to see that this issue is open for 2+ years

**Comment by mohammedayub44** [comment]
> > > @ericxsun After performing the incremental training, the model is correctly predicting with high probability for newly given values(provided in incremental training).
> > > But it does not predict correctly for other values which were working fine with the original model.
> > 
> > 
> > I am also facing the same issue. Any recommendations how to resolve this?
> 
> I'll check it and wish to solve such problem

@ericxsun  just checking if update on this ?

**Comment by phoenixbai** [comment]
it has been two years, and incr function is still not merged......

**Comment by MrRace** [comment]
Huge Sad to see that this issue has been open almost 3 years........Sad beyond words!

**Comment by abhiaj** [comment]
> training the classification model or word vector model incrementally.
> 
> ```
> ./fasttext [supervised | skipgram | cbow] -input train.data -inputModel trained.model.bin -output re-trained [other options] -incr
> ```
> 
> `-incr` stands for incremental training.
> 
> When training word embedding, one could do it from scratch with all data at each time, or just on the new data. For classification, one could train it from scratch with pre-trained word embedding with all data, or only the new one, with no changing of the word embedding.
> 
> Incremental training actually means, having finished training model with data we got before, and retrain the model with newer data we get, **not from scratch**.

What is the format of train.data here? Is it the same as the one used for training fasttext

**Comment by filipjnc** [comment]
What's the reason for this not getting merged? What can the community do?

**Comment by fmigone** [comment]
Just to be around

**Comment by loretoparisi** [comment]
This PR should be definitely merged! So that I can add it to [FastTextJS](https://github.com/loretoparisi/fasttext.js/) that also features WASM support to NodeJS (so server-side)! Let's merge!!!

**Comment by NixOTrix** [comment]
Four years in progress, would be amazing to see this functionality merged in!

**Comment by gembin** [comment]
Seems this project is no longer actively maintained

**Comment by loretoparisi** [comment]
> Seems this project is no longer actively maintained

I think right now, the best option is to use online training provided by gensim Word2vec, that supports FastText model loader as well:

```bash
>>> cap_path = datapath("crime-and-punishment.bin")
>>> fb_model = load_facebook_model(cap_path)
>>> 'computer' in fb_model.wv.key_to_index  # New word, currently out of vocab
False
>>> old_computer = np.copy(fb_model.wv['computer'])  # Calculate current vectors
>>> fb_model.build_vocab(new_sentences, update=True)
>>> fb_model.train(new_sentences, total_examples=len(new_sentences), epochs=model.epochs)
>>> new_computer = fb_model.wv['computer']
>>> np.allclose(old_computer, new_computer, atol=1e-4)  # Vector has changed, model has learnt something
False
>>> 'computer' in fb_model.wv.key_to_index  # New word is now in the vocabulary
True
```

Note the `update` flag here

```
>>> fb_model.build_vocab(new_sentences, update=True)
```

https://radimrehurek.com/gensim/models/fasttext.html

**Comment by Shashankwer** [comment]
gensim's fasttext model does not support ngram parameter more than 1. Is there any workaround so we can continue using ngram parameter with fasttext?

**Comment by SergeiAlonichau** [comment]
I found that incremental training you implemented is not working.

The training A + training B does not give anywhere close PR as training A+B .

This PR fixes all the issues: https://github.com/facebookresearch/fastText/pull/1327 or here https://github.com/SergeiAlonichau/fastText . 

It allows for:
1. test after each epoch
2. checkpointing
3. training on large data which does not fit into memory (largest I tested was 1.6TB) 
4. finetuning already trained models

The trained model is indistinguishable from a model that was created by an original tool and can be used for inference by the old code.

**Comment by loretoparisi** [comment]
> I found that incremental training you implemented is not working.
> 
> The training A + training B does not give anywhere close PR as training A+B .
> 
> This PR fixes all the issues: #1327 or here https://github.com/SergeiAlonichau/fastText .
> 
> It allows for:
> 
> 1. test after each epoch
> 2. checkpointing
> 3. training on large data which does not fit into memory (largest I tested was 1.6TB)
> 4. finetuning already trained models
> 
> The trained model is indistinguishable from a model that was created by an original tool and can be used for inference by the old code.

Amazing! So the working incremental training is in [this](https://github.com/SergeiAlonichau/fastText/tree/IncrementalTraining) branch.

**Comment by cultivater** [comment]
> 2\. checkpointing

thanks for the amazing work, may I know whether the parameter "-incr" is necessary or not? when I add it, it shows "Unknown argument: -incr"
